sql-plugin/src/main/scala/com/nvidia/spark/rapids/collectionOperations.scala

```diff
@@ -16,10 +16,74 @@
 
 package com.nvidia.spark.rapids
 
-import ai.rapids.cudf.ColumnVector
+import scala.collection.mutable.ArrayBuffer
+
+import ai.rapids.cudf.{ColumnVector, ColumnView}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
 
 import org.apache.spark.sql.catalyst.expressions.Expression
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+case class GpuConcat(children: Seq[Expression]) extends GpuComplexTypeMergingExpression {
+
+  @transient override lazy val dataType: DataType = {
+    if (children.isEmpty) {
+      StringType
+    } else {
+      super.dataType
+    }
+  }
+
+  override def nullable: Boolean = children.exists(_.nullable)
+
+  override def columnarEval(batch: ColumnarBatch): Any = dataType match {
+    case StringType => stringConcat(batch)
+    case ArrayType(_, _) => listConcat(batch)
+  }
+
+  private def stringConcat(batch: ColumnarBatch): GpuColumnVector = {
+    val rows = batch.numRows()
+
+    withResource(ArrayBuffer.empty[ColumnVector]) { buffer =>
+      withResource(Scalar.fromString(null)) { nullScalar =>
+        // build input buffer
+        children.foreach { child =>
+          child.columnarEval(batch) match {
+            case cv: GpuColumnVector =>
+              buffer += cv.getBase
+            case null =>
+              buffer += GpuColumnVector.from(nullScalar, rows, StringType).getBase
+            case sv: Any =>
+              withResource(Scalar.fromString(sv.asInstanceOf[UTF8String].toString)) { scalar =>
+                buffer += GpuColumnVector.from(scalar, rows, StringType).getBase
+              }
+          }
+        }
+        // run string concatenate
+        withResource(Scalar.fromString("")) { emptyScalar =>
+          GpuColumnVector.from(ColumnVector.stringConcatenate(emptyScalar, nullScalar,
+            buffer.toArray[ColumnView]), StringType)
+        }
+      }
+    }
+  }
+
+  private def listConcat(batch: ColumnarBatch): GpuColumnVector = {
+    withResource(ArrayBuffer[ColumnVector]()) { buffer =>
+      // build input buffer
+      children.foreach { child =>
+        child.columnarEval(batch) match {
+          case cv: GpuColumnVector => buffer += cv.getBase
+          case _ => throw new UnsupportedOperationException("Unsupported GpuScalar of List")
+        }
+      }
+      // run list concatenate
+      GpuColumnVector.from(ColumnVector.listConcatenateByRow(buffer: _*), dataType)
```

Comment:
We verify in the java code that there are at least two columns.

https://github.com/rapidsai/cudf/blob/92f067dcd80ce11c0431432f3817bebfa97816a0/java/src/main/java/ai/rapids/cudf/ColumnVector.java#L551

I don't know if this is a bug in the java code or if the underlying list concat in cudf really does need at least two columns.  Either way CPU Spark supports concat with a single column.

```
scala> spark.range(2).selectExpr("concat(array(id, id))").show
+---------------------+
|concat(array(id, id))|
+---------------------+
|               [0, 0]|
|               [1, 1]|
+---------------------+
```

If the cudf implementation supports it, then we should file an issue just like with the string support and fix it before this goes in. If not then we need to put in a work around here.

Reply:
I checked native side, both `cudf::strings::concatenate` and `cudf::lists::concatenate_rows` supports single column inputs. And I filed issues against Java side assertion:
* https://github.com/rapidsai/cudf/issues/8246
* https://github.com/NVIDIA/spark-rapids/issues/2414